src/main/GCNode.cpp

```diff
@@ -262,121 +199,81 @@ void GCNode::mark()
     GCStackRootBase::visitRoots(&marker);
     ProtectStack::visitRoots(&marker);
     WeakRef::markThru();
-    if (R_Srcref)
-	marker(R_Srcref);
-}
-
-static GCNode* getNodePointerFromAllocation(void* allocation)
-{
-    return reinterpret_cast<GCNode*>(
-	get_object_pointer_from_allocation(allocation));
-}
-
-void GCNode::detachReferentsOfObjectIfUnmarked(GCNode* object,
-					       vector<GCNode*> *unmarked_and_saturated)
-{
-    if (!object->isMarked()) {
-	int ref_count = object->getRefCount();
-	incRefCount(object);
-	if (object->getRefCount() == ref_count) {
-	    // The reference count has saturated.
-	    object->detachReferents();
-	    unmarked_and_saturated->push_back(object);
-	} else {
-	    object->detachReferents();
-	    decRefCount(object);
-	}
+    if (R_Srcref) {
+        marker(R_Srcref);
     }
 }
 
-void GCNode::sweep()
-{
+void GCNode::sweep() {
     // Detach the referents of nodes that haven't been marked.
     // Once this is done, all of the nodes in the cycle will be unreferenced
     // and they will have been deleted unless their reference count is
     // saturated.
-    vector<GCNode*> unmarked_and_saturated;
-    applyToAllAllocatedNodes([&](GCNode* node) {
-	    detachReferentsOfObjectIfUnmarked(node, &unmarked_and_saturated);
-	});
+    vector<void*> work;
+    vector<GCNode*> to_delete;
+    BlockPool::ApplyToAllBlocks([&](void* pointer) {
+        work.push_back(pointer);
```

Comment:
The worklist was used to avoid problems caused by deleting nodes while iterating over live nodes. This is not a problem in the current version of the allocator since it does not use an std::map or other datastructure that disallows concurrent iteration/modification.

The worklist could be removed right now. The only pitfall is if the bitset is not volatile, in which case the allocator could mistakenly identify something as allocated. I had this problem in yet another previous version of the allocator.